### ISSUE TITLE
[FloatingActionButton]  implementing floatingActionButton component

### ISSUE THREAD

**Issue by wndudqus** [dooboolab/dooboo-ui#63]
**Floating action button(FAB) is a very common component in touch based devices.**

[material design](https://material.io/components/buttons-floating-action-button#usage)

There are many libraries that supports floating action button, and it is good for highlights the most important action on the page.

![image](https://user-images.githubusercontent.com/35381940/128805382-a80f5b97-fb66-4300-a046-a10ed8d18f8d.png)

**I think we should support FAB.**

I'm proposing FAB that has the following features:
1. that can be placed in an absolute position on the screen.
2. that allows you to set icons with content
3. that can show additional action buttons when selected

**Comment by wndudqus** [comment]
I think it's similar to IconButton component.
But it has difference in feature 1, 3
If you decide to implement it, assign it to me.

**Comment by hyochan** [comment]
> I think it's similar to IconButton component.
> But it has difference in feature 1, 3
> If you decide to implement it, assign it to me.

Assigned 🚀

**Comment by yujong-lee** [comment]
@wndudqus 
 Looks good to me. 
Like you mentioned, I think It will be nice to get component like `Button` and `IconButton` as props.

```js
<Floating {...props}>
  <Button onClick={...}>
</Floating>
```
or for the feature 3,

```js
<Floating
  main={<Button {...props}/>}
  sub={[<IconButton name='abc'/>, <IconButon name='bcd'/>]}
>
```
How do you think?

I'm looking forward to your work. 😄

**Comment by wndudqus** [comment]
@yujong-lee  
Thank you. I was thinking in the same way

**Comment by hyochan** [comment]
> @wndudqus
> Looks good to me.
> Like you mentioned, I think It will be nice to get component like `Button` and `IconButton` as props.
> 
> ```js
> <Floating {...props}>
>   <Button onClick={...}>
> </Floating>
> ```
> 
> or for the feature 3,
> 
> ```js
> <Floating
>   {...props}
>   main={<Button onClick-{...}/>}
>   sub={[<IconButton name='abc'/>, <IconButon name='bcd'/>]}
> >
> ```
> 
> How do you think?
> 
> I'm looking forward to your work. 😄

@yujong-lee I just want to point out DRY for same elements in `sub`. Same `IconButton` should not be repeated.

**Comment by hyochan** [comment]
If someone wants to support custom element in specifix index, it can be provided like below.
```
<FAB
   renderItems={(item, i) => {
    if (i === 3) return <MyView />;
    return <View />;
  }
...
/>
```

Just my idea 😉

**Comment by wndudqus** [comment]
I think you meant this.

```TS

<FAB
   renderItems={(item, i) => {
    if (i === 3) return <MyView />;
    return item;//return original Component
  }
...
/>

```

Did I get it right?
If I get It right,I think that is wonderful Idea. Thanks!

**Comment by yujong-lee** [comment]
> > @wndudqus
> > Looks good to me.
> > Like you mentioned, I think It will be nice to get component like `Button` and `IconButton` as props.
> > ```js
> > <Floating {...props}>
> >   <Button onClick={...}>
> > </Floating>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or for the feature 3,
> > ```js
> > <Floating
> >   {...props}
> >   main={<Button onClick-{...}/>}
> >   sub={[<IconButton name='abc'/>, <IconButon name='bcd'/>]}
> > >
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How do you think?
> > I'm looking forward to your work. 😄
> 
> @yujong-lee I just want to point out DRY for same elements in `sub`. Same `IconButton` should not be repeated.

@hyochan 

Thanks for pointing it out!

I think repeating doesn't matter because I can always remove it using function returning array of components. (for simple cases `map` is sufficient). But It will end up writing render function, so eventually your approach might be better.

So now I think the point is how `renderFuction` should defined.

**Comment by hyochan** [comment]
> I think you meant this.
> 
> ```ts
> <FAB
>    renderItems={(item, i) => {
>     if (i === 3) return <MyView />;
>     return item;//return original Component
>   }
> ...
> />
> ```
> 
> Did I get it right?
> If I get It right,I think that is wonderful Idea. Thanks!

The `item` is a data not an `Element`. You might want to refer to [FlatList](https://reactnative.dev/docs/flatlist). This is a great component made by Facebook!

**Comment by wndudqus** [comment]
I created PR about this issue.
I want to talk about structure of **[this FAB](https://github.com/dooboolab/dooboo-ui/pull/67)** component.

If you have some time please check it out.
Or should I implement more specifically?

**Comment by yujong-lee** [comment]
Resolved with #67.
